cpp/pagerank_module/algorithm_online/pagerank.hpp

```diff
@@ -0,0 +1,146 @@
+  ///
+  ///
+ public:
+  void Init() {
+    walks.clear();
+    walks_counter.clear();
+    walks_table.clear();
+  }
+
+  bool IsEmpty() const { return walks.empty(); }
+
+  /// Keeping the information about walks on the graph
+  std::vector<std::vector<std::uint64_t>> walks;
+
+  // Keeping the information of walk appearance in algorithm for faster calculation
+  std::unordered_map<std::uint64_t, uint64_t> walks_counter;
+
+  /// Table that keeps the node appearance and walk ID
+  std::unordered_map<std::uint64_t, std::unordered_set<std::uint64_t>> walks_table;
+};
+
+extern PageRankData context;
+extern std::uint64_t global_R;
+extern double global_epsilon;
+
+///
+///@brief Function for vector normalization
+///
+///@param rank Vector that needs to be normalized
+///
+void NormalizeRank(std::vector<double> &rank);
+
+///
+///@brief Calculates pagerank based on current information stored in global context
+///
+///@return std::vector<std::pair<std::uint64_t, double>>
+///
+std::vector<std::pair<std::uint64_t, double>> CalculatePageRank();
+
+///
+///@brief Creates a route starting from start_id, stores it in walk and updates the walk_index. Route is created via
+/// random walk depending on random number genrator.
+///
+///@param graph Graph for route creation
+///@param start_id Starting node in graph creation
+///@param walk Walk vector that stores a route
+///@param walk_index Index of a walk for context storing
+///@param epsilon Probability of stopping the route creation
+///
+void CreateRoute(const mg_graph::GraphView<> &graph, std::uint64_t start_id, std::vector<std::uint64_t> &walk,
+                 std::uint64_t walk_index, double epsilon);
+
+///
+///@brief Updates the context based on adding the new vertex. This means adding it to a context tables and creating
+/// walks from it.
+///
+///@param graph Graph for updating
+///@param new_vertex New vertex
+///
+void UpdateCreate(const mg_graph::GraphView<> &graph, std::uint64_t new_vertex);
+
+///
+///@brief Updates the context based on new edge addition. Reverts previous walks made from starting node and updates
+/// them.
+///
+///@param graph Graph for updating
+///@param new_edge New edge
+///
+void UpdateCreate(const mg_graph::GraphView<> &graph, const std::pair<std::uint64_t, std::uint64_t> &new_edge);
+
+///
+///@brief Removes the edge from the context and updates walks. This method works by updating walks that contain starting
+/// node because they no longer exist.
+///
+///@param graph Graph for updating
+///@param removed_edge Deleted edge
+///
+void UpdateDelete(const mg_graph::GraphView<> &graph, const std::pair<std::uint64_t, std::uint64_t> &removed_edge);
+
+///
+///@brief Deletes vertex from context. This is trivial because we are sure that no edge exists around that node.
+///
+///@param graph Graph for updating
+///@param removed_vertex Removed vertex
+///
+void UpdateDelete(const mg_graph::GraphView<> &graph, std::uint64_t removed_vertex);
+
+}  // namespace
```

Comment:
This should be deleted and the docs moved above the implementations as it's not supposed to be exposed to the user.

Reply:
The pattern with anonymous namespaces is to have them only in cpp and not exposed in hpp file?

Reply:
Not only is it a pattern, but it's a huge no-no to put anonymous namespaces in header files: 
https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file
:smile: